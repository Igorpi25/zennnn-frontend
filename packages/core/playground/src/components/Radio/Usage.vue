<template>
  <div class="mb-6">
    Value: {{ model }}
  </div>
  <Radio
    v-for="(item, i) in items"
    v-model="model"
    :key="item"
    :value="item"
    :disabled="i === 2"
    name="radio-group"
    class="pb-1"
  >
    <span class="ml-2">{{ item }}</span>
  </Radio>
</template>

<script setup>
import { ref } from 'vue'

const model = ref(null)
const items = ['item-1', 'item-2', 'item-3', 'item-4']
</script>
