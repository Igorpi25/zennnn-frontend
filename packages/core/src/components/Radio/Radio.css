@layer components {
  .radio__control {
    @apply flex items-start relative;
  }
  .radio--show-details .radio__control {
    @apply pb-2;
  }
  .radio__control__input {
    @apply relative;
  }
  .radio__control__icon {
    @apply w-5 h-5 border-2 border-gray-300 dark:border-gray-300 rounded-full;
    @apply flex items-center justify-center;
    @apply transition-colors duration-100 ease-in-out;
    @apply text-white;
  }
  .radio--focused .radio__control__icon {
    @apply ring ring-blue-400;
  }
  .radio__input:focus:not(:focus-visible) + .radio__control__icon {
    box-shadow: none;
    outline: 0;
  }
  .radio--active .radio__control__icon {
    @apply border-blue-500 bg-blue-500;
  }
  .radio--active.radio--disabled .radio__control__icon {
    @apply bg-gray-300 border-gray-300 dark:bg-gray-300 dark:border-gray-300;
  }

  .radio__input {
    @apply absolute opacity-0 select-none w-full h-full;
  }

  .radio__label {
    @apply select-none leading-tight;
  }

  .radio__label,
  .radio__input {
    @apply cursor-pointer;
  }
  .radio--disabled .radio__label,
  .radio--disabled .radio__input {
    @apply cursor-not-allowed;
  }

  .radio--disabled {
    @apply opacity-50;
  }
}
