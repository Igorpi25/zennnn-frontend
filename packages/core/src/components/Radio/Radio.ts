import {
  h,
  ref,
  computed,
  watch,
  vShow,
  withDirectives,
  defineComponent,
} from 'vue'
import { deepEqual } from 'vue-supp'

import { ziStatusPoint } from '@zennnn/icons'

import { useInputProps, useInput } from '../../composables/useInput'
import { useInputValidationProps, useInputValidation } from '../../composables/useInputValidation'

import uid from '../../utils/uid'

import Icon from '../Icon'

export default defineComponent({
  name: 'Radio',

  props: {
    ...useInputProps(),
    ...useInputValidationProps(),
    modelValue: {
      type: [String, Number],
      default: '',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    valueComparator: {
      type: Function,
      default: deepEqual,
    },
  },

  emits: ['update:modelValue', 'update:error', 'change'],

  setup (props, { slots, emit }) {
    const id: string = uid('radio-')
    const rootElement = ref<HTMLElement>()

    const {
      internalValue,
      isFocused,
      genInput,
      genLabel,
    } = useInput(props, { slots, id })

    const {
      showDetails,
      isDisabled,
      isReadonly,
      genMessages,
    } = useInputValidation(props, { emit, id, internalValue, isFocused })

    const isActive = computed(() => {
      const value = props.value
      const input = internalValue.value

      return value
        ? props.valueComparator(value, input)
        : Boolean(input)
    })

    watch(() => props.modelValue, (val) => {
      internalValue.value = val
    })

    watch(internalValue, (val) => {
      emit('update:modelValue', val)
    })

    const onChange = (e: Event) => {
      if (isDisabled.value || isReadonly.value || isActive.value) return

      const target = e.target as HTMLInputElement

      const value = target.value

      internalValue.value = value
      emit('change', value)
    }

    const genRadioLabel = () => {
      const children = slots.default?.()
      if (!children) return undefined
      return h('label', {
        for: id,
        class: 'radio__label',
      }, children)
    }

    const genRadioIcon = () => {
      return h('div', {
        class: 'radio__control__icon',
      }, withDirectives(
        h(Icon, {
          size: 16,
        }, {
          default: () => ziStatusPoint,
        }),
        [
          [vShow, isActive.value],
        ],
      ))
    }

    const genRadioInput = () => {
      const data = {
        id,
        class: 'radio__input',
        value: props.value,
        checked: isActive.value,
        type: 'radio',
        role: 'radio',
        readonly: isReadonly.value,
        disabled: isDisabled.value,
        'aria-checked': isActive.value,
        onChange: onChange,
      }
      return genInput(data)
    }

    const genRadio = () => {
      return h('div', {
        class: 'radio__control__input',
      }, [
        genRadioInput(),
        genRadioIcon(),
      ])
    }

    const genControl = () => {
      return h('div', {
        class: {
          radio__control: true,
        },
      }, [
        genRadio(),
        genRadioLabel(),
      ])
    }

    return () => {
      return h('div', {
        ref: rootElement,
        class: {
          radio: true,
          'radio--active': isActive.value,
          'radio--focused': isFocused.value,
          'radio--disabled': isDisabled.value,
          'radio--readonly': isReadonly.value,
          'radio--show-details': showDetails.value,
        },
      }, [
        genLabel(),
        genControl(),
        genMessages(),
      ])
    }
  },
})
