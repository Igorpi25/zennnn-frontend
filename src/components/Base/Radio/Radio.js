import {
  h,
  ref,
  computed,
  watch,
  vShow,
  withDirectives,
} from 'vue'

import { ziCircleFill } from '../../../assets/icons'

import { useInputProps, useInput } from '../../../composables/useInput'
import { useInputValidationProps, useInputValidation } from '../../../composables/useInputValidation'

import uid from '../../../utils/uid'
import { deepEqual } from '../../../utils/deepEqual'

import Icon from '../Icon'

import './Radio.css'

export default {
  name: 'Radio',

  props: {
    ...useInputProps(),
    ...useInputValidationProps(),
    modelValue: {
      type: [String, Number],
      default: '',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    valueComparator: {
      type: Function,
      default: deepEqual,
    },
  },

  emits: ['update:modelValue', 'update:error', 'change'],

  setup (props, { slots, emit }) {
    const id = uid('radio-')
    const rootElement = ref(null)

    const {
      internalValue,
      isFocused,
      genInput,
      genLabel,
    } = useInput(props, { slots, id })

    const {
      showDetails,
      isDisabled,
      isReadonly,
      genMessages,
    } = useInputValidation(props, { emit, id, internalValue, isFocused })

    const isActive = computed(() => {
      const value = props.value
      const input = internalValue.value

      return value
        ? props.valueComparator(value, input)
        : Boolean(input)
    })

    watch(() => props.modelValue, (val) => {
      internalValue.value = val
    })

    watch(internalValue, (val) => {
      emit('update:modelValue', val)
    })

    const onChange = (e) => {
      if (isDisabled.value || isReadonly.value || isActive.value) return

      const value = e.target.value

      internalValue.value = value
      emit('change', value)
    }

    const genRadioLabel = () => {
      const children = slots.default ? slots.default() : undefined
      if (!children) return null
      return h('label', {
        for: id,
        class: 'radio__label',
      }, children)
    }

    const genRadioIcon = () => {
      return h('div', {
        class: 'radio__control__icon',
      }, withDirectives(
        h(Icon, {
          size: 16,
        }, {
          default: () => ziCircleFill,
        }),
        [
          [vShow, isActive.value],
        ],
      ))
    }

    const genRadioInput = () => {
      const data = {
        id,
        class: 'radio__input',
        value: props.value,
        checked: isActive.value,
        type: 'radio',
        role: 'radio',
        readonly: isReadonly.value,
        disabled: isDisabled.value,
        'aria-checked': isActive.value,
        onChange: onChange,
      }
      return genInput(data)
    }

    const genRadio = () => {
      return h('div', {
        class: 'radio__control__input',
      }, [
        genRadioInput(),
        genRadioIcon(),
      ])
    }

    const genControl = () => {
      return h('div', {
        class: {
          radio__control: true,
        },
      }, [
        genRadio(),
        genRadioLabel(),
      ])
    }

    return () => {
      return h('div', {
        ref: rootElement,
        class: {
          radio: true,
          'radio--active': isActive.value,
          'radio--focused': isFocused.value,
          'radio--disabled': isDisabled.value,
          'radio--readonly': isReadonly.value,
          'radio--show-details': showDetails.value,
        },
      }, [
        genLabel(),
        genControl(),
        genMessages(),
      ])
    }
  },
}
