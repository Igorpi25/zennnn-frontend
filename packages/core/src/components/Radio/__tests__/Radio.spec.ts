import { mount, VueWrapper } from '@vue/test-utils'

import Radio from '../Radio'

jest.mock('../../../utils/uid')

describe('Radio.ts', () => {
  let mountFunction: (options?: Record<string, unknown>) => VueWrapper<any>

  beforeEach(() => {
    mountFunction = (options = {}) => {
      return mount(Radio, {
        ...options,
      })
    }
  })

  // Render

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with disabled prop and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        disabled: true,
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with name, required, readonly props and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        name: 'simple-input',
        required: true,
        readonly: true,
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render checked component and match snapshot', () => {
    const wrapper = mountFunction({
      props: {
        modelValue: 'radio-1',
        value: 'radio-1',
      },
      slots: {
        default: () => 'Label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render focused component and match snapshot', async () => {
    const wrapper = mountFunction({
      slots: {
        default: () => 'Label',
      },
    })

    await wrapper.get('input').trigger('focus')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with label prop and match snapshot', async () => {
    const wrapper = mountFunction({
      props: {
        label: 'Input label',
      },
      slots: {
        default: () => 'Radio label',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // ---

  // TODO: onchange return 'on'
  it.todo('should emit value when input/label clicked')
})
