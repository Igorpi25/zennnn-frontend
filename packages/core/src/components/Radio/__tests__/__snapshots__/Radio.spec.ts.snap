// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio.ts should render checked component and match snapshot 1`] = `
<div class="radio radio--active">
  <!---->
  <div class="radio__control">
    <div class="radio__control__input"><input id="radio-1" class="radio__input" type="radio" value="radio-1" role="radio" aria-checked="true">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Radio.ts should render component and match snapshot 1`] = `
<div class="radio">
  <!---->
  <div class="radio__control">
    <div class="radio__control__input"><input id="radio-1" class="radio__input" type="radio" role="radio" aria-checked="false">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true" style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Radio.ts should render component with disabled prop and match snapshot 1`] = `
<div class="radio radio--disabled">
  <!---->
  <div class="radio__control">
    <div class="radio__control__input"><input disabled="" id="radio-1" class="radio__input" type="radio" role="radio" aria-checked="false">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true" style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Radio.ts should render component with label prop and match snapshot 1`] = `
<div class="radio"><label id="label-1" for="radio-1" aria-hidden="false" class="label">
    <div class="label__wrapper">
      <div class="label__content__wrapper">
        <div class="label__content">Input label</div>
      </div>
      <!---->
    </div>
  </label>
  <div class="radio__control">
    <div class="radio__control__input"><input id="radio-1" class="radio__input" type="radio" role="radio" aria-checked="false">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true" style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Radio label</label>
  </div>
  <!---->
</div>
`;

exports[`Radio.ts should render component with name, required, readonly props and match snapshot 1`] = `
<div class="radio radio--readonly">
  <!---->
  <div class="radio__control">
    <div class="radio__control__input"><input name="simple-input" required="" readonly="" id="radio-1" class="radio__input" type="radio" role="radio" aria-checked="false">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true" style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Label</label>
  </div>
  <!---->
</div>
`;

exports[`Radio.ts should render focused component and match snapshot 1`] = `
<div class="radio radio--focused">
  <!---->
  <div class="radio__control">
    <div class="radio__control__input"><input id="radio-1" class="radio__input" type="radio" role="radio" aria-checked="false" value="">
      <div class="radio__control__icon"><i class="icon" aria-hidden="true" style="display: none;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="16px" width="16px" role="img" aria-hidden="true" style="font-size: 16px; height: 16px; width: 16px;">
            <g fill="currentColor">
              <path d="M18 12a6 6 0 11-12 0 6 6 0 0112 0z"></path>
            </g>
          </svg></i></div>
    </div><label for="radio-1" class="radio__label">Label</label>
  </div>
  <!---->
</div>
`;
